src/logic/safe/transactions/gasNew.js

```diff
@@ -58,3 +76,75 @@ export const estimateTxGasCosts = async (
     return 10000
   }
 }
+
+export const estimateSafeTxGas = async (
+  safe: any,
+  safeAddress: string,
+  data: string,
+  to: string,
+  valueInWei: number | string,
+  operation: number,
+) => {
+  try {
+    let safeInstance = safe
+    if (!safeInstance) {
+      safeInstance = await getGnosisSafeInstanceAt(safeAddress)
+    }
+
+    const web3 = await getWeb3()
+    const estimateData = safeInstance.contract.methods.requiredTxGas(to, valueInWei, data, operation).encodeABI()
+    const estimateResponse = await web3.eth.call({
+      to: safeAddress,
+      from: safeAddress,
+      data: estimateData,
+    })
+    const txGasEstimation = new BigNumber(estimateResponse.substring(138), 16).toNumber() + 10000
+
+    // 21000 - additional gas costs (e.g. base tx costs, transfer costs)
+    const dataGasEstimation = estimateDataGasCosts(estimateData) + 21000
+
+    let additionalGasBatches = [10000, 20000, 40000, 80000, 160000, 320000, 640000, 1280000, 2560000, 5120000]
+    for (let i = 0; i < 10; i++) {
+      const batch = new web3.BatchRequest()
+
+      const estimationRequests = additionalGasBatches.map(
+        (additionalGas) =>
+          new Promise((resolve) => {
+            const request = web3.eth.call.request(
+              {
+                to: safe.address,
+                from: safe.address,
+                data: estimateData,
+                gasPrice: 0,
+                gasLimit: txGasEstimation + dataGasEstimation + additionalGas,
+              },
+              (error) => {
+                resolve({
+                  success: error ? false : true,
```

Comment:
is this equal to checking `estimateResponse != '0x'`?

Reply:
Not sure, probably worth checking the response too